src/stan/services/pathfinder/single.hpp

```diff
@@ -0,0 +1,1104 @@
+ * approximate draws, and a vector of the lp ratio. If `false`, only returns an
+ * error code `error_codes::OK` if successful, `error_codes::SOFTWARE` for
+ * failures
+ */
+template <bool ReturnLpSamples = false, class Model, typename DiagnosticWriter,
+          typename ParamWriter>
+inline auto pathfinder_lbfgs_single(
+    Model& model, const stan::io::var_context& init, unsigned int random_seed,
+    unsigned int path, double init_radius, int history_size, double init_alpha,
+    double tol_obj, double tol_rel_obj, double tol_grad, double tol_rel_grad,
+    double tol_param, int num_iterations, bool save_iterations, int refresh,
+    callbacks::interrupt& interrupt, int num_elbo_draws, int num_draws,
+    callbacks::logger& logger, callbacks::writer& init_writer,
+    ParamWriter& parameter_writer, DiagnosticWriter& diagnostic_writer) {
+  const auto start_optim_time = std::chrono::steady_clock::now();
+  boost::ecuyer1988 rng
+      = util::create_rng<boost::ecuyer1988>(random_seed, path);
+  std::vector<int> disc_vector;
+  std::vector<double> cont_vector = util::initialize<false>(
+      model, init, rng, init_radius, false, logger, init_writer);
+  const auto num_parameters = cont_vector.size();
+  // Setup LBFGS
+  std::stringstream lbfgs_ss;
+  stan::optimization::LSOptions<double> ls_opts;
+  ls_opts.alpha0 = init_alpha;
+  stan::optimization::ConvergenceOptions<double> conv_opts;
+  conv_opts.tolAbsF = tol_obj;
+  conv_opts.tolRelF = tol_rel_obj;
+  conv_opts.tolAbsGrad = tol_grad;
+  conv_opts.tolRelGrad = tol_rel_grad;
+  conv_opts.tolAbsX = tol_param;
+  conv_opts.maxIts = num_iterations;
+  using lbfgs_update_t
+      = stan::optimization::LBFGSUpdate<double, Eigen::Dynamic>;
+  lbfgs_update_t lbfgs_update(history_size);
+  using Optimizer
+      = stan::optimization::BFGSLineSearch<Model, lbfgs_update_t, double,
+                                           Eigen::Dynamic, true>;
+  Optimizer lbfgs(model, cont_vector, disc_vector, std::move(ls_opts),
+                  std::move(conv_opts), std::move(lbfgs_update), &lbfgs_ss);
+  const std::string path_num("Path: [" + std::to_string(path) + "] ");
+  if (refresh != 0) {
+    logger.info(path_num + "Initial log joint density = "
+                + std::to_string(lbfgs.logp()));
+  }
+  std::vector<std::string> names;
+  model.constrained_param_names(names, true, true);
+  names.push_back("lp_approx__");
+  names.push_back("lp__");
+  parameter_writer(names);
+  int ret = 0;
+  std::vector<Eigen::VectorXd> param_vecs;
+  param_vecs.reserve(num_iterations);
+  std::vector<Eigen::VectorXd> grad_vecs;
+  grad_vecs.reserve(num_iterations);
+  Eigen::VectorXd prev_params
+      = Eigen::Map<Eigen::VectorXd>(cont_vector.data(), cont_vector.size());
+  Eigen::VectorXd prev_grads;
+  boost::circular_buffer<Eigen::VectorXd> param_buff(history_size);
+  boost::circular_buffer<Eigen::VectorXd> grad_buff(history_size);
+  std::size_t current_history_size = 0;
+  {
```

Comment:
Personally I just like to have braces here so that it's obvious `g1` will not be used in the rest of the function. idt there is enough here to warrant a separate function, but I also want to convey that this is a silo'd thing. So the middle ground for that felt like just having braces so it's apparent what is a tmp for this scope.

Reply:
I'm not going to object too strongly, but I feel that the natural conclusion of this line of thinking is that we mark the end of scope to every variable this way and our programs wind up with more braces than Lisp!

Reply:
So feel free to just make this resolved if you plan to leave the braces.